moveit_ros/occupancy_map_monitor/src/occupancy_map_monitor.cpp

```diff
@@ -39,124 +39,95 @@
 #include <moveit_msgs/srv/load_map.hpp>
 #include <moveit/occupancy_map_monitor/occupancy_map.h>
 #include <moveit/occupancy_map_monitor/occupancy_map_monitor.h>
+#include <moveit/occupancy_map_monitor/occupancy_map_monitor_middleware_handle.hpp>
 #include <boost/bind.hpp>
 
 namespace occupancy_map_monitor
 {
-static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit.ros.occupancy_map_monitor");
-
-OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node, double map_resolution)
-  : map_resolution_(map_resolution), debug_info_(false), mesh_handle_count_(0), node_(node), active_(false)
+namespace
 {
-  initialize();
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit.ros.occupancy_map_monitor");
 }
 
-OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node,
+OccupancyMapMonitor::OccupancyMapMonitor(std::unique_ptr<MiddlewareHandle> middleware_handle,
                                          const std::shared_ptr<tf2_ros::Buffer>& tf_buffer,
                                          const std::string& map_frame, double map_resolution)
-  : tf_buffer_(tf_buffer)
+  : middleware_handle_(std::move(middleware_handle))
+  , tf_buffer_(tf_buffer)
   , map_frame_(map_frame)
   , map_resolution_(map_resolution)
   , debug_info_(false)
   , mesh_handle_count_(0)
-  , node_(node)
+  , active_(false)
 {
   initialize();
 }
 
+OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node, double map_resolution)
+  : OccupancyMapMonitor(std::make_unique<OccupancyMapMonitorMiddlewareHandle>(node), nullptr, "", map_resolution)
+{
+}
+
+OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node,
+                                         const std::shared_ptr<tf2_ros::Buffer>& tf_buffer,
+                                         const std::string& map_frame, double map_resolution)
+  : OccupancyMapMonitor(std::make_unique<OccupancyMapMonitorMiddlewareHandle>(node), tf_buffer, map_frame,
+                        map_resolution)
+{
+}
+
 void OccupancyMapMonitor::initialize()
 {
-  /* load params from param server */
-  if (map_resolution_ <= std::numeric_limits<double>::epsilon())
-    if (!node_->get_parameter("octomap_resolution", map_resolution_))
-    {
-      map_resolution_ = 0.1;
-      RCLCPP_WARN(LOGGER, "Resolution not specified for Octomap. Assuming resolution = %g instead", map_resolution_);
-    }
+  map_resolution_ = middleware_handle_->getMapResolutionParameter();
   RCLCPP_DEBUG(LOGGER, "Using resolution = %lf m for building octomap", map_resolution_);
 
   if (map_frame_.empty())
-    if (!node_->get_parameter("octomap_frame", map_frame_))
-      if (tf_buffer_)
-        RCLCPP_WARN(LOGGER, "No target frame specified for Octomap. No transforms will be applied to received data.");
+  {
+    map_frame_ = middleware_handle_->getMapFrameParameter();
+  }
 
+  if (tf_buffer_ && map_frame_.empty())
```

Comment:
Nit: comparing pointers explicitly to `nullptr`s is clearer than using the overload.

Reply:
Just putting it out there (not trying to change things), this is an area where cppcoreguidelines and abseil disagree

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es87-dont-add-redundant--or--to-conditions

https://abseil.io/tips/141